### PR TITLE
feat!: Implement plugin as LSP client-only

### LINKED ISSUE #47: Lowercase drive letter on Windows
VSCode seems to provide file paths with a lowercase drive letter (actually, it's even more weird, `file:///c:/Users/User/...`) and we do a bunch of normalization in order for the Grain compiler to work. For Packaged Grain, it fails if the drive letter is lowercase, so in #45 I needed to add a normalization that capitalized it. We should figure out the more proper fix.

### LINKED ISSUE #92: Handle multi root folder workspaces properly for compile script
VSCode supports multiple root level folders in a workspace.  

The vscode server implementation doesn't yet support determining on the server side which root folder a file is found in.

This  is more obvious where the folders are in distinct paths but not when the folders are inside each other.

I'd suggest sorting the root folders by length and then matching the URI of the file against the most specific folder URI to determine the best matching folder.

You then need to track per folder if there's a build script, and to use it to compile if it exists.  We currently use chokidar to monitor these files, so that monitoring would need to be reset

There is an event sent when the workspace folders change to allow this to be recalculated.

### PR DESCRIPTION
**Pull request by marcusroberts**
Updated extension to work with the new long running LSP server process

With the updated LSP protocol we can remove the hacks needed for handling lens updates, making the LSP protocol implementation match the standard.

To be released in conjunction with the updated compiler side

The server wrapper is also no longer needed as the client communicates directly with the LSP process.

Closes #47 because we do URI normalization in the `grain lsp` CLI
Closes #92 because we start an LSP per top-level workspace & each file not inside a workspace gets its own LSP instance.

### CODE REVIEW

**Comment by marcusroberts** [comment]
This needs a fix to run on Windows (looking for grain.cmd) and I also need to restore the use of the build script for setting custom flags

**Comment by marcusroberts** [comment]
Handles using .cmd on Windows and looking for the grainfind.js build script

File: editor-extensions/vscode/package.json [hunk 1]
```diff
@@ -2,15 +2,15 @@
   "name": "vscode-grain",
   "displayName": "Grain",
   "description": "Grain support for Visual Studio Code.",
-  "version": "0.15.0",
+  "version": "0.16.0",
```
Comment (phated) [inline editor-extensions/vscode/package.json:5]: release-please handles bumping this, correct?

File: editor-extensions/vscode/package.json [hunk 2]
```diff
@@ -69,25 +75,25 @@
...
         },
-        "grain_language_server.enableStatementLenses": {
+        "grain.enableStatementLenses": {
           "scope": "window",
           "type": "boolean",
           "default": true,
           "description": "Enable per-statement lenses"
         },
-        "grain_language_server.maxNumberOfProblems": {
+        "grain.maxNumberOfProblems": {
```
Comment (phated) [inline editor-extensions/vscode/package.json:90]: @marcusroberts Is this used? How/where?

File: editor-extensions/vscode/package.json [hunk 3]
```diff
@@ -69,25 +75,25 @@
...
       "properties": {
-        "grain_language_server.enableLSP": {
+        "grain.enableLSP": {
           "scope": "window",
           "type": "boolean",
           "default": true,
           "description": "Enable the language server"
         },
-        "grain_language_server.enableStatementLenses": {
+        "grain.enableStatementLenses": {
```
Comment (phated) [inline editor-extensions/vscode/package.json:84]: @marcusroberts This option doesn't actually disable lenses anymore. Do we need to supply it somewhere? Does it send a message to the LSP server?

File: editor-extensions/vscode/package.json [hunk 4]
```diff
@@ -69,25 +75,25 @@
...
           "scope": "window",
           "type": "boolean",
           "default": true,
           "description": "Enable per-statement lenses"
         },
-        "grain_language_server.maxNumberOfProblems": {
+        "grain.maxNumberOfProblems": {
           "scope": "resource",
           "type": "number",
           "default": 100,
```
Comment (phated) [inline editor-extensions/vscode/package.json:93]: We only show 1 error, right?

File: editor-extensions/vscode/README-LSP.md [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
 ## Functionality
 
-This Language Server works for grain files. It has the following language features:
+This Language Server Protocol client works for grain files. It has the following language features:
 
-- Calls grain compiler to get report type errors
+- Show compilation warnings and errors
+- Code lenses
+- Hover support
+- Code completion
```
Comment (phated) [inline editor-extensions/vscode/README-LSP.md:12]:
```suggestion
```
Remove this if we are punting it

File: editor-extensions/vscode/package.json (hunk 4 above)
Comment (marcusroberts) [reply editor-extensions/vscode/package.json]: Yes, option removed

File: editor-extensions/vscode/package.json (hunk 3 above)
Comment (marcusroberts) [reply editor-extensions/vscode/package.json]: It was there when performance was slow, removed now.

File: editor-extensions/vscode/package.json (hunk 2 above)
Comment (marcusroberts) [reply editor-extensions/vscode/package.json]: Not used any more.

File: editor-extensions/vscode/package.json (hunk 1 above)
Comment (marcusroberts) [reply editor-extensions/vscode/package.json:5]: I'm not sure, paging @ospencer

Comment (phated) [reply editor-extensions/vscode/package.json:5]:
```suggestion
  "version": "0.15.0",
```

**Review by phated — APPROVED** [review]
I'm super happy with how this is operating in combination with the new LSP. 🎉

**Review by ospencer — APPROVED** [review]
Looks great!

**Pull request merged**